### PR TITLE
Next.js image optimization fixes

### LINKED ISSUE #6135: Firebase CLI deploy of next.js app failed with error for missing file client-reference-manifest.js
### [REQUIRED] Environment info

**firebase-tools:** 12.4.4

**Platform:** macOS 13.4.1 (c) 

### [REQUIRED] Test case

I have uploaded the project creating from the steps below to GitHub at [jdstanhope/broken-thing](https://github.com/jdstanhope/broken-thing) 

### [REQUIRED] Steps to reproduce

Create a React app with using the next framework using the following selections

```
npx create-next-app

✔ What is your project named? … XXX
✔ Would you like to use TypeScript? … No / *Yes*
✔ Would you like to use ESLint? … No / *Yes*
✔ Would you like to use Tailwind CSS? … No / *Yes*
✔ Would you like to use `src/` directory? … No / *Yes*
✔ Would you like to use App Router? (recommended) … No / *Yes*
✔ Would you like to customize the default import alias? … *No* / Yes

```

Setup firebase hosting for the app

```
firebase init hosting

/path/to/home/directory/XXX

=== Project Setup

First, let's associate this project directory with a Firebase project.
You can create multiple project aliases by running firebase use --add,
but for now we'll just set up a default project.

? Please select an option: Create a new project
i  If you want to create a project in a Google Cloud organization or folder, please use "firebase projects:create" instead, and return to this command when you've created the project.
? Please specify a unique project id (warning: cannot be modified afterward) [6-30 characters]:
 XXX
? What would you like to call your project? (defaults to your project ID)
✔ Creating Google Cloud Platform project
✔ Adding Firebase resources to Google Cloud Platform project

🎉🎉🎉 Your Firebase project is ready! 🎉🎉🎉

Project information:
   - Project ID: XXX
   - Project Name: XXX

=== Hosting Setup
? Detected an existing Next.js codebase in the current directory, should we use this? Yes
? In which region would you like to host server-side content, if applicable? us-central1 (Iowa)
? Set up automatic builds and deploys with GitHub? No

i  Writing configuration info to firebase.json...
i  Writing project information to .firebaserc...

✔  Firebase initialization complete!
``` 

Deploy the app to firebase using

```
firebase deploy --only hosting
```

### [REQUIRED] Expected behavior

App is deployed without an error.

### [REQUIRED] Actual behavior

App deployment failed with an error in the firebase-debug.log with the following

```
[debug] [2023-07-16T18:13:25.722Z] Error: ENOENT: no such file or directory, open '/path/to/project/XXX/.next/server/client-reference-manifest.js'
```

### PR DESCRIPTION
**Pull request by leoortizz**
### Description

- Fixes #6135
- Fix `next/image` component in app directory for Next.js > 13.4.9.
- Require a Cloud Function when Next.js Image Optimization is used in the app directory.
- Refactor `getNextVersion` to prevent issues with canary versions,  add unit tests.

### CODE REVIEW

File: src/frameworks/next/utils.ts [hunk 1]
```diff
@@ -209,25 +212,80 @@ export async function isUsingMiddleware(dir: string, isDevMode: boolean): Promis
...
+
+  // Note: casting the types here as this function is called after the Next.js build is done,
+  // so Next.js was already identified as a dependency. In this case getNextVersion
+  // will always return a valid string
+  const nextVersion = getNextVersion(projectDir) as string;
+  const nextVersionSemver = minVersion(nextVersion) as SemVer;
+
+  // Note: canary Next.js versions e.g. 13.4.10-canary.0 are identified as lower than 13.4.10 by SemVer.
+  // Here we use only the major, minor and patch versions for comparison.
+  const nextVersionString = `${nextVersionSemver.major}.${nextVersionSemver.minor}.${nextVersionSemver.patch}`;
```
Comment (alexastrum) [inline src/frameworks/next/utils.ts:257]: Create a util method if one does not exist already.
Pretty sure I saw similar logic in other places.

`getNextReleaseVersion(projectDir)`

File: src/frameworks/next/utils.ts [hunk 2]
```diff
@@ -209,25 +212,80 @@ export async function isUsingMiddleware(dir: string, isDevMode: boolean): Promis
...
+      if (fileIncludesNextImageComponent(fileContents.toString())) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  // Next.js < 13.4.10 has a single client-reference-manifest.js file
+  const clientReferenceManifest = await readFile(
```
Comment (alexastrum) [inline src/frameworks/next/utils.ts:282]: Move this logic (`Next.js < 13.4.10`) above the more complex `Next.js >= 13.4.10` logic.
So it's in chronological order and from simple to complex.

File: src/frameworks/next/utils.ts [hunk 3]
```diff
@@ -209,25 +212,80 @@ export async function isUsingMiddleware(dir: string, isDevMode: boolean): Promis
...
+
+  // Note: canary Next.js versions e.g. 13.4.10-canary.0 are identified as lower than 13.4.10 by SemVer.
+  // Here we use only the major, minor and patch versions for comparison.
+  const nextVersionString = `${nextVersionSemver.major}.${nextVersionSemver.minor}.${nextVersionSemver.patch}`;
+
+  if (gte(nextVersionString, "13.4.10")) {
+    // Next.js >= 13.4.10 has one client-reference-manifest file per entry
+    const files = await new Promise<string[]>((resolve) => {
+      glob(
+        join(nextDir, "server", "app", "**", "*client-reference-manifest.js"),
```
Comment (alexastrum) [inline src/frameworks/next/utils.ts:263]: Wouldn't `join(nextDir, "server", "**", "*client-reference-manifest.js")` work for all Nextjs version?

File: src/frameworks/next/utils.ts [hunk 4]
```diff
@@ -209,25 +212,80 @@ export async function isUsingMiddleware(dir: string, isDevMode: boolean): Promis
...
 }
 
+/**
+ * Whether next/image is being used in the app directory
+ */
+async function isUsingNextImageInAppDirectory(
+  projectDir: string,
+  nextDir: string
+): Promise<boolean> {
+  function fileIncludesNextImageComponent(fileContents: string): boolean {
```
Comment (alexastrum) [inline src/frameworks/next/utils.ts:245]: rename to `includesNextImageComponent(fileContents: string)`

File: src/test/frameworks/next/utils.spec.ts [hunk 5]
```diff
@@ -245,22 +245,22 @@ describe("Next.js utils", () => {
       stub.withArgs(EXPORT_MARKER).resolves(exportMarkerWithImage);
       stub.withArgs(IMAGES_MANIFEST).resolves(imagesManifest);
 
-      expect(await isUsingImageOptimization("")).to.be.true;
```
Comment (alexastrum) [inline src/test/frameworks/next/utils.spec.ts:248]: Add 2 tests for `*client-reference-manifest.js`.
One for legacy, and another one for v`13.4.10` containing app dir.

**Review by alexastrum — CHANGES_REQUESTED** [review]

File: src/frameworks/next/utils.ts (hunk 2 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: good call, thanks!

File: src/frameworks/next/utils.ts (hunk 1 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: I found a way better way to do this using `semver`'s `coerce` , so I moved this logic to `getNextVersion` and added unit tests for that

File: src/test/frameworks/next/utils.spec.ts (hunk 5 above)
Comment (leoortizz) [reply src/test/frameworks/next/utils.spec.ts:248]: Added these test cases to the `isUsingNextImageInAppDirectory` tests instead as that's the function handling `client-reference-manifest.js`

File: src/frameworks/next/utils.ts (hunk 3 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: Good call too, thanks! I removed all the Next versions comparisons and used just this glob

**Review by alexastrum — APPROVED** [review]

**Pull request merged**